File: practice/algorithms/implementation/sherlock_and_squares/SherlockAndSquares.js
'use strict';

const processData = input => {
    const lines = input.split('\n').slice(1);
    for(let i = 0; i < lines.length; i++) {
        const parts = lines[i].split(' ');
        const lower = parts[0];
        const upper = parts[1];
        let squareNumbers = 0;
        if(lower === upper) {
            if (Math.sqrt(lower) === Math.ceil(Math.sqrt(lower))) {
                squareNumbers = 1;
            }
            else {
                squareNumbers = 0;
            }
        }
        else {
            squareNumbers = parseInt(Math.floor(Math.sqrt(upper)) - (Math.ceil(Math.sqrt(lower))) + 1)
        }
        console.log(squareNumbers);
    }
};

process.stdin.resume();
process.stdin.setEncoding("ascii");

let _input = "";
process.stdin.on("data", input => _input += input);
process.stdin.on("end", () => processData(_input));

File: practice/algorithms/implementation/sherlock_and_squares/SherlockAndSquares.scala
object SherlockAndSquares extends App {
  val tests = Source.stdin.getLines().drop(1)
  val numberIntervals = tests.map(line => line.splitAt(line.indexOf(" ")))
    .map(tuple => (tuple._1.trim.toInt, tuple._2.trim.toInt))
  val squareNumbers = numberIntervals.map(getSquareNumbers)
  println(squareNumbers.mkString("\n"))

  def getSquareNumbers(interval: (Int, Int)): Int = {
    val lower = interval._1
    val upper = interval._2
    if(lower == upper)
      if(Math.sqrt(lower) == Math.ceil(Math.sqrt(lower)))
        1
      else
        0
    else
      (Math.floor(Math.sqrt(upper)) - (Math.ceil(Math.sqrt(lower))) + 1).toInt
  }
}
